### PR TITLE
Replaces Big Mob Tiny Sprites With Togglable Clientside Transparency For Big Mobs

### PR DESCRIPTION
**Pull request by IndieanaJones**
## About The Pull Request

This PR replaces the tiny mob sprite abilities for our large mobs with a Toggle Seethrough ability. What this does when enabled is make yourself transparent on your own client and makes you unable to click yourself. With this, any sort of overlays that get added look correct while in your tiny mob form and you can still appreciate your large size without giving up the ability to click on targets that would be hiding underneath your sprite.

XenoQueen

image

In total, this ability is granted to all the mobs who had tiny sprites:

- Xenomorph Queen
- Space Dragon
- Megafauna (not that players ever get access to these normally, but it might be nice for admins)

Along with all this, I've also gone and expanded the mobs who get access to the Seethrough ability to mobs that are large but didn't have a mini-sprite. This includes:

- Jungle giants (leaper and mega arachnid)
- Tree mob
- Festivus pole mob
- Lavaland elites
- Xenomorph Praetorian
- Wumborian Fugu

On top of all this, any simple/basic mob given a fugu gland also automatically gains this ability.

This solution also fixes all the overlays bugs that tiny mob sprites introduced to the game, most notably the xenomorph queen overlays being screwy when in small form and space dragon's overlays being completely disabled for everyone if he was in small form.

The only issue with this currently is that clicking on any overlays from your mob will let you still click on yourself, though this issue is pretty negligible.

## Why It's Good For The Game

Tiny mob sprite abilities clashed very hard with our overlay system, leading to a lot of buggy and unintended behavior. This replacement works a lot more smoothly with overlays and also requires much less setup to give it to a mob than tiny sprites did. I would also imagine it helps keep the fantasy of being a big monster alive by letting you still see yourself being a big monster without it getting in your way when trying to attack things.

## Changelog
:cl:
del: Removed tiny mob sprite abilities for large mobs
add: Add a Self Seethrough ability, given to most large mobs and all mobs affected by a fugu gland
fix: Space Dragon's overlay will no longer vanish at times

### CODE REVIEW

File: code/modules/mob/living/basic/festivus_pole.dm [hunk 1]
```diff
@@ -42,6 +42,8 @@
 
 /mob/living/basic/festivus/Initialize(mapload)
 	. = ..()
+	var/datum/action/toggle_seethrough/seeThrough = new
+	seeThrough.Grant(src)
```
Comment (MrMelbert) [inline code/modules/mob/living/basic/festivus_pole.dm:46]: 1. Instead of having the action give the component, just give the component to the mob, and have the component create the action. Makes it a bit easier to make use of your component. 

```dm
/mob/whatever/Initialize(mapload)
    . = ..()
    AddComponent(/datum/component/seethrough_mob)
```
```dm
/datum/component/seethrough_mob/Initialize(...)
    . = ..()
    var/datum/action/toggle_seethrough/action= new(src)
    action.Grant(parent)
```

File: code/datums/components/seethrough_mob.dm [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	. = ..()
+
+/datum/action/toggle_seethrough/Trigger(trigger_flags)
+	..()
+	var/datum/component/seethrough_mob/seethroughComp = owner.GetComponent(/datum/component/seethrough_mob)
+	seethroughComp.is_active = !seethroughComp.is_active
+	if(seethroughComp.is_active)
+		seethroughComp.trick_mob()
+	else
+		seethroughComp.untrick_mob()
```
Comment (MrMelbert) [inline code/datums/components/seethrough_mob.dm:123]: 2. This allows for us to create the action with the target = the component, so instead of doing a GetComponent, we can just do `var/datum/component/seethrough_mob/seethroughComp = target`.

Comment (MrMelbert) [inline code/datums/components/seethrough_mob.dm:123]: 3. Then we can move all of the logic into the component to keep this a bit contained. 
```dm
var/datum/component/seethrough_mob/seethroughComp = target
seethroughComp.toggle_active()
```
```dm
/datum/component/seethrough_mob/proc/toggle_active()
	is_active = !is_active
	if(is_active)
		trick_mob()
	else
		untrick_mob()
```

File: code/datums/components/seethrough_mob.dm [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		grant_to.AddComponent(/datum/component/seethrough_mob)
+
+/datum/action/toggle_seethrough/Remove(mob/remove_from)
+	var/datum/component/seethrough_mob/seethroughComp = owner.GetComponent(/datum/component/seethrough_mob)
+	if(seethroughComp.is_active)
+		seethroughComp.untrick_mob()
+	. = ..()
+
+/datum/action/toggle_seethrough/Trigger(trigger_flags)
+	..()
```
Comment (MrMelbert) [inline code/datums/components/seethrough_mob.dm:117]: Don't forget to respect parent return value for `Trigger`.

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by IndieanaJones** [comment]
Originally, I didn't do it like you suggested as I was concerned with how fugu would work and getting two versions of the component, but testing the changes in-game seems to have the second component instance not do anything. So yeah, that works better and now admins can slap it on whatever they want plus it being one line of code as opposed to two.

**Comment by nukularpower** [comment]
just wanted to say, i might be misunderstanding, but you said the clicking on yourself is "negligable" - i 100% disagree, clicking on yourself is the main reason i use these small forms, and I would not want to see them removed if this didnt solve that particular issue personally

**Comment by IndieanaJones** [comment]
> just wanted to say, i might be misunderstanding, but you said the clicking on yourself is a "negligible" issue - i 100% disagree, clicking on yourself is the main reason i use these small forms, and I would not want to see them removed if this didnt solve that particular issue personally

You're misunderstanding. The clear form of the mob is 100% unclickable, but any overlays (fire overlays on xeno queen, for example) will let you hit yourself if you click on them. This isn't desired behavior, but it was the same for small mob sprites and when a solution is found to resolve that issue, it will be implemented.

**Comment by MrMelbert** [comment]
> I didn't do it like you suggested as I was concerned with how fugu would work and getting two versions of the component

In case you weren't aware, by default components have a dupe mode of `COMPONENT_DUPE_HIGHLANDER`, which - if a second component of the same type is added to a datum - the old component will be deleted and the new component will be added, leaving one component.

I would actually recommend you use `COMPONENT_DUPE_UNIQUE`, which does the opposite - If you add a second component, the new component will simply not be added and leave the old component there.

File: code/datums/components/seethrough_mob.dm [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	RegisterSignal(fool, COMSIG_MOB_LOGOUT, PROC_REF(on_client_disconnect))
+
+///Remove the screen object and make us appear solid to ourselves again
+/datum/component/seethrough_mob/proc/untrick_mob()
+	var/mob/fool = parent
+	animate(trickery_image, alpha = 255, time = animation_time)
+	UnregisterSignal(fool, COMSIG_MOB_LOGOUT)
+
+	//after playing the fade-in animation, remove the image and the trick atom
+	addtimer(CALLBACK(src, TYPE_PROC_REF(/datum/component/seethrough,clear_image), trickery_image, fool.client), animation_time)
```
Comment (MrMelbert) [inline code/datums/components/seethrough_mob.dm:78]: You don't need to use `TYPE_PROC_REF` if it's for src
```suggestion
	addtimer(CALLBACK(src, PROC_REF(clear_image), trickery_image, fool.client), animation_time)
```

File: code/datums/components/seethrough_mob.dm [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+/datum/action/toggle_seethrough/Remove(mob/remove_from)
+	var/datum/component/seethrough_mob/seethroughComp = target
+	if(seethroughComp.is_active)
+		seethroughComp.untrick_mob()
+	. = ..()
+
+/datum/action/toggle_seethrough/Trigger(trigger_flags)
+	if(!..())
+		return
```
Comment (MrMelbert) [inline code/datums/components/seethrough_mob.dm:121]:
```suggestion
	. = ..()
	if(!.)
		return
```
Just so we end up returning `TRUE` anyways.

File: code/datums/components/seethrough_mob.dm [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	desc = "Allows you to see behind your massive body and click through it."
+	button_icon = 'icons/mob/actions/actions_xeno.dmi'
+	button_icon_state = "alien_sneak"
+	background_icon_state = "bg_alien"
+
+/datum/action/toggle_seethrough/Remove(mob/remove_from)
+	var/datum/component/seethrough_mob/seethroughComp = target
+	if(seethroughComp.is_active)
+		seethroughComp.untrick_mob()
+	. = ..()
```
Comment (MrMelbert) [inline code/datums/components/seethrough_mob.dm:117]:
```suggestion
	return ..()
```

**Comment by itseasytosee** [comment]
Are there any reasons these mobs might want to click on themselves?

**Comment by IndieanaJones** [comment]
> Are there any reasons these mobs might want to click on themselves?

No, I don't believe so. Clicking on yourself just hits yourself on all of them.

**Comment by Serxule** [comment]
the only thing i dont like is that this is still harder to see with than having a 1 tile sprite even if you cant click on yourself

**Comment by IndieanaJones** [comment]
> it also has the problem of some big mobs that you have no idea where you're actually standing, which was a huge upside of the tiny sprite that you actually knew where you were

Considering xeno queen, lesser ash drake and space dragon were the only playables who had access to this ability, which one did you have issue figuring out where you were standing with?

**Comment by Serxule** [comment]
> > it also has the problem of some big mobs that you have no idea where you're actually standing, which was a huge upside of the tiny sprite that you actually knew where you were
> 
> Considering xeno queen, lesser ash drake and space dragon were the only playables who had access to this ability, which one did you have issue figuring out where you were standing with?

any of them, especially when moving, its just harder to know exactly where you are when your sprite takes up 4+ tiles

**Comment by IndieanaJones** [comment]
> any of them, especially when moving, its just harder to know exactly where you are when your sprite takes up 4+ tiles

Unfortunately, I can't really do much to help you out there then, as I want the small sprites gone for the reasons listed in the PR body. Best I can hope for is being able to use the new system will help you figure out where the big mobs exist on their tile after playing with it long enough.

I would consider adding a config option in order for the system to be flexible for people of your opinion, but any form of smaller sprite will end up breaking the overlays.

File: code/datums/components/seethrough_mob.dm [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	var/mob/fool = parent
+	var/datum/hud/our_hud = fool.hud_used
+	for(var/atom/movable/screen/plane_master/seethrough in our_hud.get_true_plane_masters(SEETHROUGH_PLANE))
+		seethrough.unhide_plane(fool)
+
+	render_source_atom = new()
+
+	var/static/uid = 0
+	uid++
+	fool.render_target = "*transparent_bigmob[uid]"
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:42]: Mhm... Is there's no other way that doesn't override the mob's render target? Doing it this way is guaranteed to get into the way of many other features.
It'd be nice if we could have a middleman that uses the mob's render target as source and this value as render target.

File: code/datums/components/seethrough_mob.dm [hunk 8]
```diff
@@ -0,0 +1,124 @@
+///A component that lets you turn your character transparent in order to see and click through yourself.
+/datum/component/seethrough_mob
+	///The atom that enables our dark magic
+	var/obj/render_source_atom
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:4]:
```suggestion
	var/atom/movable/render_source_atom
```
less variables, more lightweight (disregard in case of unknown variables).

File: code/datums/components/seethrough_mob.dm [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+///Set up everything we need to trick the client and keep it looking normal for everyone else
+/datum/component/seethrough_mob/proc/trick_mob()
+	SIGNAL_HANDLER
+
+	var/mob/fool = parent
+	var/datum/hud/our_hud = fool.hud_used
+	for(var/atom/movable/screen/plane_master/seethrough in our_hud.get_true_plane_masters(SEETHROUGH_PLANE))
+		seethrough.unhide_plane(fool)
+
+	render_source_atom = new()
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:38]: It would be less wasteful to instantiate and delete this once, instead of everytime.

File: code/datums/components/seethrough_mob.dm [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+	//after playing the fade-in animation, remove the image and the trick atom
+	addtimer(CALLBACK(src, PROC_REF(clear_image), trickery_image, fool.client), animation_time)
+
+///Remove the image and the trick atom
+/datum/component/seethrough_mob/proc/clear_image(image/removee, client/remove_from)
+	var/atom/atom_parent = parent
+	atom_parent.appearance_flags &= ~KEEP_TOGETHER
+	atom_parent.render_target = null
+	qdel(render_source_atom)
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:85]: ditto

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/datums/components/seethrough_mob.dm (hunk 7 above)
Comment (IndieanaJones) [reply code/datums/components/seethrough_mob.dm:42]: This whole system relies on * being the leading value in order to make the base render of the mob invisible, and ultimately there is no way around it. Potato did some thinking on it and there wasn't really an alternate solution without setting up a whole backend for handling cases like this one, and he told me ultimately I could PR it with the current solution and it would be corrected at some later time.

File: code/datums/components/seethrough_mob.dm (hunk 8 above)
Comment (IndieanaJones) [reply code/datums/components/seethrough_mob.dm:4]: The game got upset with me when I tried sending a atom/movable to nullspace.

File: code/datums/components/seethrough_mob.dm (hunk 10 above)
Comment (IndieanaJones) [reply code/datums/components/seethrough_mob.dm:85]: Can you rephrase that?

File: code/datums/components/seethrough_mob.dm (hunk 7 above)
Comment (Ghommie) [reply code/datums/components/seethrough_mob.dm:42]: Well then, at least implement a way to revert the render_target change when the seethrough is removed, please.

File: code/datums/components/seethrough_mob.dm (hunk 8 above)
Comment (Ghommie) [reply code/datums/components/seethrough_mob.dm:4]: What gives, any more detail on this?

Comment (IndieanaJones) [reply code/datums/components/seethrough_mob.dm:4]: ...However, testing it right now, it seemed fine this time. I probably tried atom instead of atom/movable or something.

File: code/datums/components/seethrough_mob.dm (hunk 10 above)
Comment (Ghommie) [reply code/datums/components/seethrough_mob.dm:85]: Whoops! Just make sure `render_source_atom` is deleted when the component is also deleted.

File: code/datums/components/seethrough_mob.dm [hunk 11]
```diff
@@ -23,9 +27,18 @@
...
+	src.render_source_atom = new()
 	var/datum/action/toggle_seethrough/action = new(src)
+	var/static/uid = 0
+	uid++
+	src.personal_uid = uid
+
 	action.Grant(parent)
 
+/datum/component/seethrough_mob/Destroy(force, silent)
+	qdel(render_source_atom)
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:39]:
```suggestion
	QDEL_NULL(render_source_atom)
```

File: code/datums/components/seethrough_mob.dm [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+	var/mob/fool = parent
+	animate(trickery_image, alpha = 255, time = animation_time)
+	UnregisterSignal(fool, COMSIG_MOB_LOGOUT)
+
+	//after playing the fade-in animation, remove the image and the trick atom
+	addtimer(CALLBACK(src, PROC_REF(clear_image), trickery_image, fool.client), animation_time)
+
+///Remove the image and the trick atom
+/datum/component/seethrough_mob/proc/clear_image(image/removee, client/remove_from)
+	var/atom/atom_parent = parent
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:92]:
```suggestion
	var/atom/movable/atom_parent = parent
	atom_parent.vis_contents -= render_source_atom
```

File: code/datums/components/seethrough_mob.dm [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+	QDEL_NULL(render_source_atom)
+	return ..()
+
+///Set up everything we need to trick the client and keep it looking normal for everyone else
+/datum/component/seethrough_mob/proc/trick_mob()
+	SIGNAL_HANDLER
+
+	var/mob/fool = parent
+	var/datum/hud/our_hud = fool.hud_used
+	for(var/atom/movable/screen/plane_master/seethrough in our_hud.get_true_plane_masters(SEETHROUGH_PLANE))
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:48]:
```suggestion
	for(var/atom/movable/screen/plane_master/seethrough as anything in our_hud.get_true_plane_masters(SEETHROUGH_PLANE))
```
`as anything` should be used if we're sure that every item in a list is of the same type.

File: code/datums/components/seethrough_mob.dm [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	remove_from?.images -= removee
+
+///Effect is disabled when they log out because client gets deleted
+/datum/component/seethrough_mob/proc/on_client_disconnect()
+	SIGNAL_HANDLER
+
+	var/mob/fool = parent
+	UnregisterSignal(fool, COMSIG_MOB_LOGOUT)
+	var/datum/hud/our_hud = fool.hud_used
+	for(var/atom/movable/screen/plane_master/seethrough in our_hud.get_true_plane_masters(SEETHROUGH_PLANE))
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:103]:
```suggestion
	for(var/atom/movable/screen/plane_master/seethrough as anything in our_hud.get_true_plane_masters(SEETHROUGH_PLANE))
```

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/datums/components/seethrough_mob.dm [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+	trickery_image = new(render_source_atom)
+	trickery_image.loc = render_source_atom
+	trickery_image.override = TRUE
+
+	trickery_image.pixel_x = 0
+	trickery_image.pixel_y = 0
+
+	if(clickthrough)
+		//Special plane so we can click through the overlay
+		SET_PLANE_EXPLICIT(trickery_image, SEETHROUGH_PLANE, fool)
```
[the remark this answers is not in the document]
Comment (IndieanaJones) [reply code/datums/components/seethrough_mob.dm:73]: We can't make the image in Initialize() or the image will be invisible when we try to show it to the client. The stuff regarding the render_source_atom is able to be moved however and I did so.

We could generate one if we don't have one already and don't if we don't, but I don't know how stable that ends up being. The image stuff seems a bit finicky.

[the remark this answers is not in the document]
Comment (Ghommie) [reply code/datums/components/seethrough_mob.dm]: Ok, but what about the render source movable?

[the remark this answers is not in the document]
Comment (IndieanaJones) [reply code/datums/components/seethrough_mob.dm]: You mean putting render_source_atom.render_source = "*transparent_bigmob[personal_uid]" into Init so we don't set it every time when it doesn't change? Yeah we can do that.

[the remark this answers is not in the document]
Comment (Ghommie) [reply code/datums/components/seethrough_mob.dm]: Also the vis flags and appearance flags.

File: code/datums/components/seethrough_mob.dm [hunk 16]
```diff
@@ -41,6 +41,7 @@
 	var/icon/current_mob_icon = icon(fool.icon, fool.icon_state)
 	render_source_atom.pixel_x = -fool.pixel_x
 	render_source_atom.pixel_y = ((current_mob_icon.Height() - 32) * 0.5)
```
Comment (Ghommie) [inline code/datums/components/seethrough_mob.dm:43]: I think I've already said this, but these lines can be moved back to `trick_mob()` (that should keep the offsets a bit more up to date until dedicated signals are implemented), sorry for the inconvenience.

**Review by Ghommie — APPROVED** [review]

**Comment by IndieanaJones** [comment]
I hate merge conflicts

**Comment by Ghommie** [comment]
Forgot to remove the extra tags, eheh.
